KAFKA-14619: KRaft validate snapshot id are at batch boundries

kevin-wu24 opened the pull request:
When `KafkaRaftClient` receives a request to create a snapshot with offset that is not aligned to a batch boundary, do not create a misaligned snapshot and instead log at info level and throw an `IllegalArgumentException`.
* Perform a log read using `snapshotId.offset()` and check whether the offset in question is the base offset of the returned batch. 

Testing
* Modified unit tests in `KafkaRaftClientSnapshotTest`: Add cases in `testCreateSnapshotAsLeader/FollowerWithInvalidSnapshotId` where the snapshot offset is only not batch aligned, but otherwise valid, and modify existing cases so their snapshot offset is batch aligned
* Added tests to `KafkaMetadataLogTest` and `MockLogTest`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kevin-wu24 committed:
prevent raft client from snapshotting when snapshotid offset is not batch algined

kevin-wu24 committed:
adding all batch aligned offsets when appending records

kevin-wu24 committed:
cleanup last commit

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -258,7 +260,8 @@ public KafkaRaftClient(
             new Random(),
-            quorumConfig
+            quorumConfig,
+            new HashSet<>()
```
KRaft should not track the starting offsets in this way. There are many issues with this approach. The more important ones are:

1. This doesn't track batch start offsets that were added to the log prior to starting the KRaft replica. It also doesn't track batch start offsets that were added by the leader.
2. This doesn't handle log truncation.
3. It is very memory inefficient. It is possible for KRaft to have a lot of small batches and each one of them would have an entry in this set.

The log layer already track these offsets. We shouldn't duplicate this information in the KRaft layer. Take a look at the `KafkaMetadataLog#read` method to see how the log layer computes the batch start offset. Specifically, you can take a look at `LogSegment#translateOffset`.

kevin-wu24 replied:
Looks like the log layer gets the batch associated with a `targetOffset` from `FileRecords#searchForOffsetWithSize`, but it currently returns the last offset included in that batch. It looks safe to change this to return the base offset instead since the offset field is never being accessed, so we can retrieve the batch start offset for a given offset. However, `LogSegment#read` also needs to construct `LogOffsetMetadata` with this batch start offset instead of the `startOffset` that is passed into the method so we can retrieve this batch start offset, but this ends up breaking quite a few tests.

jsancio reviewed (commented):
Thanks for the improvement @kevin-wu24 . I have one high-level comment.

kevin-wu24 committed:
will check fetchInfo offset metadata returned from log layer to see if a snapshot offset is batch aligned

kevin-wu24 committed:
correcting last commit, want to get the base offset of a batch being read from the log for given offset

kevin-wu24 committed:
checking fetched records for batch alignment in createSnapshot

kevin-wu24 committed:
cleanup of last commit

kevin-wu24 committed:
reverting SnapshotWriterReaderTest changes

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3426,6 +3428,12 @@ public Optional<SnapshotWriter<T>> createSnapshot(
             throw new IllegalStateException("Cannot create snapshot before the replica has been initialized");
         }
 
+        AbstractIterator<? extends RecordBatch> batchIterator = log.read(snapshotId.offset(), Isolation.COMMITTED).records.batchIterator();
+        if (batchIterator.hasNext() && batchIterator.peek().baseOffset() != snapshotId.offset()) {
+            logger.error("Cannot create snapshot at offset {} because it is not batch aligned", snapshotId.offset());
+            return Optional.empty();
+        }
```
I know that I agreed to this solution in our offline discussion but since we are planning to change the UnifiedLog implementation to return the batch start offset in the LogOffsetMetadata, then maybe it is better to fix that PR (https://github.com/apache/kafka/pull/17528) first and merge it. After that we can fix this PR and Jira using the changes in the other PR.

kevin-wu24 replied:
Agreed. Just checking the offset when it actually returns the batch start offset is more readable.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3426,6 +3428,12 @@ public Optional<SnapshotWriter<T>> createSnapshot(
 
+        AbstractIterator<? extends RecordBatch> batchIterator = log.read(snapshotId.offset(), Isolation.COMMITTED).records.batchIterator();
+        if (batchIterator.hasNext() && batchIterator.peek().baseOffset() != snapshotId.offset()) {
+            logger.error("Cannot create snapshot at offset {} because it is not batch aligned", snapshotId.offset());
```
This is not an error in the KRaft layer. This is an error in the metadata layer. You can log this at info since it should be rare. In the metadata layer we can log an error for this case.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3426,6 +3428,12 @@ public Optional<SnapshotWriter<T>> createSnapshot(
+        AbstractIterator<? extends RecordBatch> batchIterator = log.read(snapshotId.offset(), Isolation.COMMITTED).records.batchIterator();
+        if (batchIterator.hasNext() && batchIterator.peek().baseOffset() != snapshotId.offset()) {
+            logger.error("Cannot create snapshot at offset {} because it is not batch aligned", snapshotId.offset());
+            return Optional.empty();
```
Instead of returning a `Optional.empty()`. This method can throw an `IllegalArgumentException` since this is a programming error in the metadata layer.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1916,7 +1916,8 @@ public void testCreateSnapshotAsLeaderWithInvalidSnapshotId(boolean withKip853Rp
         assertEquals(OptionalLong.empty(), context.client.highWatermark());
-        assertThrows(IllegalArgumentException.class, () -> context.client.createSnapshot(invalidSnapshotId1, 0));
+        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> context.client.createSnapshot(invalidSnapshotId1, 0));
+        assertEquals("Cannot create a snapshot with an id (OffsetAndEpoch(offset=4, epoch=2)) greater than the high-watermark (0)", exception.getMessage());
```
We use the following indentation in the raft module:
```java
        IllegalArgumentException exception = assertThrows(
            IllegalArgumentException.class,
            () -> context.client.createSnapshot(invalidSnapshotId1, 0)
        );
        assertEquals(
            String.format(
                "Cannot create a snapshot with an id (%s) greater than the high-watermark (%s)",
                ...
            ),
            exception.getMessage()
        );
```

This comment applies to a few places in this file.

kevin-wu24 committed:
Merge branch 'trunk' into KAFKA-14619

kevin-wu24 committed:
code review

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3426,6 +3426,12 @@ public Optional<SnapshotWriter<T>> createSnapshot(
 
+        long baseOffset = log.read(snapshotId.offset(), Isolation.COMMITTED).startOffsetMetadata.offset();
+        if (snapshotId.offset() != baseOffset) {
+            logger.info("Cannot create snapshot at offset {} because it is not batch aligned", snapshotId.offset());
```
We should also record the `baseOffset` returned by `log.read` in the log message.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3426,6 +3426,12 @@ public Optional<SnapshotWriter<T>> createSnapshot(
+        long baseOffset = log.read(snapshotId.offset(), Isolation.COMMITTED).startOffsetMetadata.offset();
+        if (snapshotId.offset() != baseOffset) {
+            logger.info("Cannot create snapshot at offset {} because it is not batch aligned", snapshotId.offset());
+            throw new IllegalArgumentException("Cannot create snapshot at an offset that is not batch aligned");
```
Let's include as much information as possible. That includes the snapshot id's end offset and the return batch's base offset?

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3426,6 +3426,12 @@ public Optional<SnapshotWriter<T>> createSnapshot(
             throw new IllegalStateException("Cannot create snapshot before the replica has been initialized");
         }
 
+        long baseOffset = log.read(snapshotId.offset(), Isolation.COMMITTED).startOffsetMetadata.offset();
```
Let's document why you are doing this and why it is correct. It may not be obvious to future readers why the snapshot's end offset must be batch aligned or point to the log end offset.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1927,18 +1934,51 @@ public void testCreateSnapshotAsLeaderWithInvalidSnapshotId(boolean withKip853Rp
 
-        OffsetAndEpoch invalidSnapshotId2 = new OffsetAndEpoch(context.client.highWatermark().getAsLong() + 2, currentEpoch);
-        assertThrows(IllegalArgumentException.class, () -> context.client.createSnapshot(invalidSnapshotId2, 0));
+        OffsetAndEpoch invalidSnapshotId2 = new OffsetAndEpoch(context.client.highWatermark().getAsLong() + 3, currentEpoch);
```
Did you need to change this because `HWM + 2` was not batched aligned since the batch after the HWM has three records? If so, I would make this explicit by using `HWM + newRecords.size()`.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1927,18 +1934,51 @@ public void testCreateSnapshotAsLeaderWithInvalidSnapshotId(boolean withKip853Rp
+            exception.getMessage()
+        );
+
+        // 4 snapshotId offset must be at a batch boundary
```
I would mention that this points to the middle of batch (d, e, f). It points to the f offset which is not batch aligned, right?

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1951,6 +1991,7 @@ public void testCreateSnapshotAsFollowerWithInvalidSnapshotId(boolean withKip853
         Set<Integer> voters = Set.of(localId, leaderId, otherFollowerId);
 
         RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters)
+            .appendToLog(0, List.of("a"))
```
Technically, `0` is not a valid leader epoch. The smallest possible epoch for a leader is 1.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -2003,11 +2054,43 @@ public void testCreateSnapshotAsFollowerWithInvalidSnapshotId(boolean withKip853
+        );
+
+        // 5) The snapshotId should be batch-aligned
+        endOffsetForEpoch = context.log.endOffsetForEpoch(currentEpoch);
+        assertEquals(4, endOffsetForEpoch.epoch());
+        assertEquals(6, endOffsetForEpoch.offset());
```
I found this difficult to follow. I assume this point to `f` in the batch `d, e, f`. That has an base offset of 3, epoch of 4 and size of 3. So you are trying to check that offset 5 (f) is not batch aligned.

The part that makes it difficult to follow is that `currentEpoch >= 5`.

The other part that make it diffcult to follow is that you are using currentEpoch in `invalidSnapshotId5` and not 4.

I think it easier to read to to make 4 a variable `batch2Epoch` and using it here and in `context.buildBatch(3L, 4, records2)`. What do you think?

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3426,6 +3426,12 @@ public Optional<SnapshotWriter<T>> createSnapshot(
             throw new IllegalStateException("Cannot create snapshot before the replica has been initialized");
         }
 
+        long baseOffset = log.read(snapshotId.offset(), Isolation.COMMITTED).startOffsetMetadata.offset();
```
Is there a reason why you added this check here and not to `ReplicatedLog` (`MockLog` and `KafkaMetadataLog`)? Those types already do snapshot id validation against the underlying log records.

kevin-wu24 replied:
Looking at this again, it makes more sense to have this moved to the `ReplicatedLog` level.

jsancio reviewed (commented):
Thanks for the changes @kevin-wu24 . Looks good in general. Here are some comments.

kevin-wu24 committed:
Merge branch 'trunk' into KAFKA-14619

kevin-wu24 committed:
code review fixes

kevin-wu24 commented:
> Thanks for the changes @kevin-wu24 . Looks good in general. Here are some comments.

Thanks for the review @jsancio. Pushed a commit:
* Now, we make the log read at the `ReplicatedLog` level, after we perform other checks in `createNewSnapshot`.
* Added/modified tests in `KafkaMetadataLogTest` and `MockLogTest`. 
* Addressed `KafkaRaftClientSnapshotTest` comments.

jsancio commented on core/src/test/scala/kafka/raft/KafkaMetadataLogTest.scala:
```diff
@@ -206,7 +217,7 @@ final class KafkaMetadataLogTest {
     val log = buildMetadataLog(tempDir, mockTime)
 
-    append(log, numberOfRecords, epoch)
+    1 to numberOfRecords foreach(_ => append(log, 1, epoch))
```
I have noticed that in Kafka we try to avoid this Scala syntax. How about
```scala
    (1 to numberOfRecords).foreach(_ => append(log, 1, epoch))
```

jsancio reviewed (commented):
LGTM. Just a minor suggestion.

kevin-wu24 committed:
code review

jsancio reviewed (approved):
LGTM
